test/Service/MailServiceTest.php

```diff
@@ -0,0 +1,238 @@
+            $this->fileSystem->url() . '/data/mail/attachments/testXlsAttachment.xls',
+            'spreadsheetName'
+        );
+
+        $result = $this->mailService->attachFiles();
+        $this->assertInstanceOf(Message::class, $result);
+        $this->assertSame('Test Subject', $result->getSubject());
+        $this->assertCount(2, $this->mailService->getAttachments());
+        $this->assertArrayHasKey('spreadsheetName', $this->mailService->getAttachments());
+    }
+
+    public function testAttachFilesToMimeMessageBody(): void
+    {
+        $stringMessage = '<div>
+            <h1>Message header</h1>
+            <div>
+                <p>Message body</p>
+            </div>
+        </div>';
+
+        $mimeMessage = new MimeMessage();
+        $mimeMessage->setParts([new Part($stringMessage)]);
+        $this->mailService->setSubject('Test Subject');
+        $this->message->setBody($mimeMessage);
+        $this->mailService->addAttachments([
+            $this->fileSystem->url() . '/data/mail/attachments/testPdfAttachment.pdf',
+            $this->fileSystem->url() . '/data/mail/attachments/testXlsAttachment.xls',
+        ]);
+
+        $result = $this->mailService->attachFiles();
+        $this->assertInstanceOf(Message::class, $result);
+        $this->assertSame('Test Subject', $result->getSubject());
+    }
+
+    public function testSendFailureTriggersErrorEvent(): void
+    {
+        $exception = new RuntimeException("Test Error Message");
+        $this->transportInterface->expects(self::once())
+            ->method('send')
+            ->willThrowException($exception);
+
+        $this->expectException(MailException::class);
+        $this->expectExceptionMessage("Test Error Message");
+        $this->mailService->send();
+    }
+
+    public function testMailResultCreatedFromException(): void
+    {
+        $customException = new RuntimeException('Custom exception test');
+
+        $mailResult = $this->mailService->createMailResultFromException($customException);
+
+        $this->assertInstanceOf(MailResult::class, $mailResult);
+        $this->assertSame($customException, $mailResult->getException());
+        $this->assertSame('Custom exception test', $mailResult->getMessage());
+    }
+
+    /**
+     * @throws Exception
+     */
+    public function testGetFolderNames(): void
+    {
+        $childFolder = $this->createMock(Folder::class);
+        $childFolder->expects(self::once())
+            ->method('getGlobalName')
+            ->willReturn('rootFolderName.childFolderName');
+
+        $rootFolder = $this->createMock(Folder::class);
+        $rootFolder->expects(self::once())
+            ->method('getChildren')
+            ->willReturn([$childFolder]);
+
+        $rootFolder->expects(self::once())
+            ->method('getGlobalName')
+            ->willReturn('rootFolderName');
+
+        $storage = $this->createMock(Imap::class);
+        $storage->expects(self::exactly(2))
```

Comment:
`willReturnOnConsecutiveCalls` is deprecated so instead of doing this:
```
$storage->expects(self::exactly(2))
    ->method('getFolders')
    ->willReturnOnConsecutiveCalls([$rootFolder], $rootFolder);
```
you should do this:
```
$storage->method('getFolders')->willReturnCallback(function () use (&$tree) {
    return array_shift($tree);
});
```